src/hooks/useSimulateLiquidity.ts

```diff
@@ -0,0 +1,161 @@
+import type {
+	AssetInfoV2,
+	LiquidityProvisionSimulation,
+	LiquidityProvisionType,
+} from "@ston-fi/api";
+import { useQuery } from "@tanstack/react-query";
+import { simulateLiquidityProvision } from "../lib/liquidity";
+import { formatAmount, toTokenUnits } from "../lib/utils";
+import { useDebounce } from "./useDebounce";
+
+/**
+ * Types for the input to the simulation hook
+ */
+export interface UseSimulateLiquidityParams {
+	walletAddress?: string;
+	/**
+	 * Full token objects so we can access addresses and decimals.
+	 */
+	tokenA?: AssetInfoV2;
+	tokenB?: AssetInfoV2;
+	amountA?: string;
+	amountB?: string;
+	provisionType: LiquidityProvisionType;
+	poolAddress?: string;
+	/**
+	 * For Balanced, we only pass one side's amount. This is determined by activeSide.
+	 */
+	activeSide?: "A" | "B" | null;
+}
+
+/**
+ * Hook to simulate liquidity provisioning with debounced inputs.
+ * Uses early returns and pure helper functions.
+ */
+export function useSimulateLiquidity(params: UseSimulateLiquidityParams) {
+	const {
+		walletAddress,
+		tokenA,
+		tokenB,
+		amountA,
+		amountB,
+		provisionType,
+		poolAddress,
+		activeSide,
+	} = params;
+
+	// Debounce user inputs
+	const debouncedAmountA = useDebounce(amountA);
+	const debouncedAmountB = useDebounce(amountB);
+	const debouncedSide = useDebounce(activeSide);
+
+	// Determine if the simulation can run
+	const isEnabled = Boolean(
+		walletAddress &&
+			tokenA?.contractAddress &&
+			tokenB?.contractAddress &&
+			debouncedAmountA &&
+			debouncedAmountB,
+	);
+
+	return useQuery<
+		LiquidityProvisionSimulation | null,
+		Error,
+		LiquidityProvisionSimulation | null,
+		readonly (string | number | undefined | null)[]
+	>({
+		queryKey: [
+			"simulateLiquidity",
+			walletAddress,
+			tokenA?.contractAddress,
+			tokenB?.contractAddress,
+			debouncedAmountA,
+			debouncedAmountB,
+			provisionType,
+			poolAddress,
+			debouncedSide,
+		],
+		queryFn: async () => {
+			if (!isEnabled) {
+				return null;
```

Comment:
I would suggest using [skipToken](https://tanstack.com/query/latest/docs/framework/react/guides/disabling-queries#typesafe-disabling-of-queries-using-skiptoken) instead of returning null in this case. This also will help to solve the TS error that forced you to use non-null asumption for tokenA & tokenB